src/project/internal/recentprojectscontroller.cpp

```diff
@@ -0,0 +1,209 @@
+ */
+#include "recentprojectscontroller.h"
+
+#include <QtConcurrent>
+
+#include <QJsonArray>
+#include <QJsonDocument>
+
+#include "async/async.h"
+#include "defer.h"
+
+using namespace mu::project;
+using namespace mu::async;
+
+void RecentProjectsController::init()
+{
+    m_dirty = true;
+
+    configuration()->rawRecentProjectsDataChanged().onNotify(this, [this]() {
+        if (!m_reloadingBlocked) {
+            m_dirty = true;
+        }
+
+        m_recentListChanged.notify();
+    });
+}
+
+const RecentFilesList& RecentProjectsController::recentProjectsList() const
+{
+    TRACEFUNC;
+
+    if (m_dirty) {
+        const_cast<RecentProjectsController*>(this)->loadRecentProjectsList();
+    } else {
+        const_cast<RecentProjectsController*>(this)->removeNonexistentFiles();
+    }
+
+    return m_recentProjectsList;
+}
+
+Notification RecentProjectsController::recentProjectsListChanged() const
+{
+    return m_recentListChanged;
+}
+
+void RecentProjectsController::prependRecentFile(const RecentFile& newFile)
+{
+    if (newFile.empty()) {
+        return;
+    }
+
+    TRACEFUNC;
+
+    RecentFilesList newList;
+    newList.reserve(m_recentProjectsList.size() + 1);
+    newList.push_back(newFile);
+
+    for (const RecentFile& file : m_recentProjectsList) {
+        if (file != newFile && fileSystem()->exists(file)) {
+            newList.push_back(file);
+        }
+    }
+
+    setRecentProjectsList(newList, true);
+
+    prependPlatformRecentFile(newFile);
+}
+
+void RecentProjectsController::clearRecentFiles()
+{
+    configuration()->setRawRecentProjectData({});
+
+    clearPlatformRecentFiles();
+}
+
+void RecentProjectsController::prependPlatformRecentFile(const io::path_t&) {}
+
+void RecentProjectsController::clearPlatformRecentFiles() {}
+
+void RecentProjectsController::loadRecentProjectsList()
+{
+    ByteArray data = configuration()->rawRecentProjectsData();
+
+    RecentFilesList newList;
+
+    DEFER {
+        setRecentProjectsList(newList, false);
+    };
+
+    if (data.empty()) {
+        return;
+    }
+
+    QJsonParseError err;
+    QJsonDocument jsonDoc = QJsonDocument::fromJson(data.toQByteArrayNoCopy(), &err);
+    if (err.error != QJsonParseError::NoError) {
+        return;
+    }
+
+    if (!jsonDoc.isArray()) {
+        return;
+    }
+
+    for (const QJsonValue val : jsonDoc.array()) {
+        newList.push_back(val.toString());
+    }
+}
+
+void RecentProjectsController::removeNonexistentFiles()
+{
+    bool removed = false;
+
+    RecentFilesList newList;
+    newList.reserve(m_recentProjectsList.size());
+
+    for (const RecentFile& file : m_recentProjectsList) {
+        if (fileSystem()->exists(file)) {
+            newList.push_back(file);
+        } else {
+            removed = true;
+        }
+    }
+
+    if (removed) {
+        setRecentProjectsList(newList, false);
+
+        async::Async::call(nullptr, [this, newList]() {
```

Comment:
Could you explain why Async::call is here?

Reply:
Just because it was also there in the old implementation in https://github.com/musescore/MuseScore/pull/17708/files#diff-77556efcf4d03b759672fe2472516dbe8c02b08208b23f69efb0865089bddbcdL133-L139 :)